Api 5355: Turn eslint rules on

anthony-kopczyk opened the pull request:
Jira Ticket: https://vajira.max.gov/browse/API-5355

This turns on the various eslint rules that were turned off when updating the backend dependencies.

Note that this makes the environment variable `SUPPORT_EMAIL` required.

anthony-kopczyk2 committed:
Update all dependencies to wanted versions

I ran npm outdated and npm update to get our packages updated to the
latest 'wanted' versions.

API-5279

anthony-kopczyk2 committed:
Remove lodash pick dependency

While updating dependencies I saw lodash pick is only used in two spots.
Since most lodash functions can be accomplished with pure ES6 syntax I
refactored the code using pick so we can remove lodash pick. That said,
unfortunately, the pick method doesn't transfer super nicely because of
how we're using it, but the code is still clean and readable.

Note: We could have made the code a little cleaner, but since the user
dynamo object is saved in dynamo, and we do not want extra content
saved, we had to be a little more declarative about how we pull out the
user fields.

API-5279

anthony-kopczyk2 committed:
Update morgan dependency to latest

API-5279

anthony-kopczyk2 committed:
Update express dependency to latest

API-5279

anthony-kopczyk2 committed:
Update typescript dependency to latest

API-5279

anthony-kopczyk2 committed:
Update axios dependency to latest

API-5279

anthony-kopczyk2 committed:
Update ts-node dependency to latest

API-5279

anthony-kopczyk2 committed:
Update sentry/node to 6.2.0 (latest)

API-5279

anthony-kopczyk2 committed:
Update @types/node dependency to 14.14.31 (latest)

API-5279

anthony-kopczyk2 committed:
Update jest related dependencies to latest

- jest to 26.6.3
- @types/jest to 26.0.20
- ts-jest to 26.5.1

API-5279

anthony-kopczyk2 committed:
Update aws-sdk to latest

API-5279

anthony-kopczyk2 committed:
Update supertest to 6.1.3 (latest)

API-5279

anthony-kopczyk2 committed:
Upgrade eslint and plugins to next major versions

Upgraded:
- eslint to 7.0.0
- @typescript-eslint/eslint-plugin to 3.0.0
- @typescript-eslint/parser to 3.0.0

Note: I added .eslintrc.js to the eslintignore file. With the major
change for parser it began giving an error that eslintrc.js was not
included in the project to be linted.

API-5279

anthony-kopczyk2 committed:
Update eslint packages to latest

Updated
- eslint to 7.20.0
- @typescript-eslint/eslint-plugin to 4.15.2
- @typescript-eslint/parser to 4.15.2

API-5279

anthony-kopczyk2 committed:
Remove extra eslint rule

There was an extra rule (that isn't even valid) in the configuration.

API-5279

anthony-kopczyk2 committed:
Run npm audit fix for security vulnerabilities

API-5279

anthony-kopczyk2 committed:
Add dotenv-config to eslintignore

dotenv-config can not understand import statements. For now it's added
to the eslintignore, but we could maybe make a separate project for
files like it.

API-5355

anthony-kopczyk2 committed:
Resolve restric-plus-operands eslint rule

This eslint rule requires the plus operand be used only with strings and
numbers, but we were using it with an error object. To resolve this I am
concatenating the string with JSON.stringify(error) so that two strings
are being concatenated.

API-5255

anthony-kopczyk2 committed:
Resolve typescript ban-types eslint errors

This eslint rule checks that we do not use specific types. For example,
object or {}. This commit resolves these errors, mostly by replacing the
types with Record<string, unknown>, which is the suggest way to resolve
this issue

API-5355

anthony-kopczyk2 committed:
Remove unneeded, commented lines

anthony-kopczyk2 committed:
Resolve error when saving to dynamo

For whatever reason, in typescript, objects of a certain type of
interface can not be passed as arguments on functions requiring the type
Record<string, unknown>. To resolve this, I modified the interface
UserDynamoeItem to be a type instead.

API-5355

anthony-kopczyk2 committed:
Resolve eslint no-unsafe-call rule failures

This resolves the errors in linting from code not respecting the
no-unsafe-call rule.

As a part of this I had to add an okta-sdk.d.ts file because there are
no type definitions. As a part of this, I moved some of our okta related
definitions into this .d.ts file because they were needed to define the
okta types properly.

API-5355

anthony-kopczyk2 committed:
Fix broken tests

While making eslint changes a few things broken.

Notably, the logging middleware in app.ts had the argument 'next'
removed which breaks the middleware. Express uses some kind of runtime
typechecking and without that argument it gets detected as the wrong
type of middleware.

This also resolves some mismatching type errors in mock implementation
callback functions in the DynamoService test.

API-5355

anthony-kopczyk2 committed:
Resolve no-unsafe-member-access eslint issues

This resolves all no-unsafe-member-access eslint problems. This required
typing many objects and casting objects in tests to the type we needed.
There may be some cleanup work to do overall with where we decide to put
different interfaces and classes.

API-5355

anthony-kopczyk2 committed:
Resolve no-unasfe-return eslint errors

This resolves all eslint errors related to having unsafe return values

API-5355

anthony-kopczyk2 committed:
Resolve restrict-template-expression eslint errors

This resolves all eslint errors in template strings where were passing
any, undefined, or other potentially invalid types to template strings.

API-5355

anthony-kopczyk2 committed:
Resolve various typescript compile errors

When we upgraded typescript this broke various tests. This resolves
those issues affecting the tests.

API-5279

anthony-kopczyk2 committed:
Merge branch 'API-5279/dependency-updates' into API-5355/turn-eslint-rules-on

anthony-kopczyk2 committed:
Use string interpolation only when necessary

Since we added conditions to check when environment variables are
undefined within our tests, we can remove the string interpolation we
added to resolve typing errors, where environment variables could be
undefined.

API-5355

anthony-kopczyk2 committed:
Resolve broken tests

The Okta mocks were trying to use the previous Okta model file that is
now within the okta-sdk.d.ts file - so updated that to import from the
okta sdk package.

Updated HealthCheck.integration to check additional environment
variables are defined before attempting to use them.

API-5355

anthony-kopczyk2 committed:
Resolve no-unsafe-assignment eslint errors

This should resolve all 'no-unsafe-assignment' eslint errors present in
the code. It also removes the eslint rule.

API-5355

anthony-kopczyk2 committed:
Bump aws-sdk, nock, eslint versions

API-5279

anthony-kopczyk2 committed:
Merge branch 'API-5279/dependency-updates' into API-5355/turn-eslint-rules-on

anthony-kopczyk2 committed:
End files in new lines

anthony-kopczyk2 committed:
Clarify comment

API-5355

anthony-kopczyk2 committed:
Merge branch 'master' into API-5355/turn-eslint-rules-on

johanna-adhoc committed:
Merge branch 'master' into API-5355/turn-eslint-rules-on

anthony-kopczyk2 committed:
Initialize OKTA_ORG to blank string

When configuring Okta, we have a condition that checks all necessary
environment variables are defined. If they are not, it throws an error.
This works, but after the code that throws the error, the linter is not
able to detect that OKTA_ORG must defined in the scenario we are using
it. Because it thinks it might be undefined, we were getting a linting
error when using it in a string template expression.

This was originally resolved by casting it to a string, but as a better
solution, we are initializing it to an empty string. This gives us a
string type and is also falsy, so our condition checking for the
variables to be defined will still function appropriately.

API-5355

anthony-kopczyk2 committed:
Merge get-environment and environments utils

The environments file exported some constants while the get-environment
util exported a function that used those constants. This commit just
merges those two files into one

API-5355

anthony-kopczyk2 committed:
Merge branch 'API-5355/turn-eslint-rules-on' of github.com:department-of-veterans-affairs/developer-portal-backend into API-5355/turn-eslint-rules-on

anthony-kopczyk2 committed:
Remove unneeded .env.test variable

I accidentally duplicated a part of the environment file while working
with the eslint rules. This removes that duplicated variable

API-5355

johanna-adhoc reviewed (dismissed):
Passed manual testing of all the endpoints in my local. LGTM.

anthony-kopczyk2 committed:
Fix broken tests

I reverted a change to the .env.test file in an earlier commit that
caused the GovDeliveryService test to fail. This reverts a similar
commit to the GovDeliveryService so it will pass again.

API-5355

TS000 reviewed (approved):
All tests pass both on local and in the container. Nice work!
